### ISSUE TITLE
Pull request #699 breaks any data download with proxies - gaierror: [Errno 11001] getaddrinfo failed

### ISSUE THREAD

**Issue by jslorrma** [gboeing/osmnx#751]
**Problem description**
Pull request #699 breaks any data download in case proxies are used or have to be used, e.g. if one is using osmnx from within a corporate network.  

Reason is that `socket.gethostbyname` can only resolve the URL's domain to an IP address for hosts within the corporate network. Trying to resolve an external URL raises the following error: `gaierror: [Errno 11001] getaddrinfo failed`

**Provide a complete minimal reproducible example**

```python
import osmnx as ox

lami, lama, lomi, loma = (18.505, 18.550, 73.83, 73.870)
G=ox.graph_from_bbox(lami, lama, lomi, loma, network_type="drive")
```

that gives the following error:

```python
~\.conda\envs\ox\lib\site-packages\osmnx\graph.py in graph_from_bbox(north, south, east, west, network_type, simplify, retain_all, truncate_by_edge, clean_periphery, custom_filter)
     86         truncate_by_edge=truncate_by_edge,
     87         clean_periphery=clean_periphery,
---> 88         custom_filter=custom_filter,
     89     )
     90 

~\.conda\envs\ox\lib\site-packages\osmnx\graph.py in graph_from_polygon(polygon, network_type, simplify, retain_all, truncate_by_edge, clean_periphery, custom_filter)
    418 
    419         # download the network data from OSM within buffered polygon
--> 420         response_jsons = downloader._osm_network_download(poly_buff, network_type, custom_filter)
    421 
    422         # create buffered graph from the downloaded data

~\.conda\envs\ox\lib\site-packages\osmnx\downloader.py in _osm_network_download(polygon, network_type, custom_filter)
    506     for polygon_coord_str in polygon_coord_strs:
    507         query_str = f"{overpass_settings};(way{osm_filter}(poly:'{polygon_coord_str}');>;);out;"
--> 508         response_json = overpass_request(data={"data": query_str})
    509         response_jsons.append(response_json)
    510     utils.log(

~\.conda\envs\ox\lib\site-packages\osmnx\downloader.py in overpass_request(data, pause, error_pause)
    699 
    700     # resolve url to same IP even if there is server round-robin redirecting
--> 701     _config_dns(base_endpoint)
    702 
    703     # define the Overpass API URL, then construct a GET-style URL as a string to

~\.conda\envs\ox\lib\site-packages\osmnx\downloader.py in _config_dns(url)
    278     """
    279     host = urlparse(url).netloc.split(":")[0]
--> 280     ip = socket.gethostbyname(host)
    281 
    282     def _getaddrinfo(*args):

gaierror: [Errno 11001] getaddrinfo failed
```

**Environment information**
-  conda version : 4.8.4
- conda-build version : 3.20.2
- python version : 3.7.11
- osmnx version : 1.1.1

<details>

```

# packages in environment at ~\.conda\envs\ox:
#
# Name                    Version                   Build  Channel
affine                    2.3.0                      py_0    defaults
argcomplete               1.12.3             pyhd3eb1b0_0    defaults
argon2-cffi               20.1.0           py37h2bbff1b_1    defaults
async_generator           1.10             py37h28b3542_0    defaults
attrs                     21.2.0             pyhd3eb1b0_0    defaults
aws-c-common              0.4.57               ha925a31_1    defaults
aws-c-event-stream        0.1.6                hd77b12b_5    defaults
aws-checksums             0.1.9                ha925a31_0    defaults
aws-sdk-cpp               1.8.185              hd77b12b_0    defaults
backcall                  0.2.0              pyhd3eb1b0_0    defaults
blas                      1.0                         mkl    defaults
bleach                    4.0.0              pyhd3eb1b0_0    defaults
bosch-ca                  1.0                           1    defaults
bottleneck                1.3.2            py37h2a96729_1    defaults
branca                    0.4.2              pyhd8ed1ab_0    conda-forge
brotli                    1.0.9                ha925a31_2    defaults
brotlipy                  0.7.0           py37h2bbff1b_1003    defaults
bzip2                     1.0.8                he774522_0    defaults
ca-certificates           2021.5.30            h5b45459_0    conda-forge
certifi                   2021.5.30        py37h03978a9_0    conda-forge
cffi                      1.14.6           py37h2bbff1b_0    defaults
cfitsio                   3.470                he774522_6    defaults
charset-normalizer        2.0.4              pyhd3eb1b0_0    defaults
click                     7.1.2              pyhd3eb1b0_0    defaults
click-plugins             1.1.1                      py_0    defaults
cligj                     0.7.2            py37haa95532_0    defaults
colorama                  0.4.4              pyhd3eb1b0_0    defaults
cryptography              3.4.7            py37h71e12ea_0    defaults
curl                      7.78.0               h86230a5_0    defaults
cycler                    0.10.0                   py37_0    defaults
debugpy                   1.4.1            py37hd77b12b_0    defaults
decorator                 5.0.9              pyhd3eb1b0_0    defaults
defusedxml                0.7.1              pyhd3eb1b0_0    defaults
entrypoints               0.3                      py37_0    defaults
expat                     2.4.1                h6c2663c_2    defaults
fiona                     1.8.13.post1     py37hd760492_0    defaults
folium                    0.12.0             pyhd8ed1ab_1    conda-forge
fonttools                 4.25.0             pyhd3eb1b0_0    defaults
freetype                  2.10.4               hd328e21_0    defaults
freexl                    1.0.6                h2bbff1b_0    defaults
gdal                      3.0.2            py37hb978731_1    defaults
geopandas                 0.9.0                      py_1    defaults
geopandas-base            0.9.0                      py_1    defaults
geos                      3.8.0                h33f27b4_0    defaults
geotiff                   1.6.0                h5770a2b_0    defaults
hdf4                      4.2.13               h712560f_2    defaults
hdf5                      1.10.6               h7ebc959_0    defaults
icc_rt                    2019.0.0             h0cc432a_1    defaults
icu                       58.2                 ha925a31_3    defaults
idna                      3.2                pyhd3eb1b0_0    defaults
importlib-metadata        3.10.0           py37haa95532_0    defaults
importlib_metadata        3.10.0               hd3eb1b0_0    defaults
intel-openmp              2021.3.0          haa95532_3372    defaults
ipykernel                 6.2.0            py37haa95532_1    defaults
ipython                   7.26.0           py37hd4e2768_0    defaults
ipython_genutils          0.2.0              pyhd3eb1b0_1    defaults
ipywidgets                7.6.3              pyhd3eb1b0_1    defaults
jedi                      0.18.0           py37haa95532_1    defaults
jinja2                    3.0.1              pyhd3eb1b0_0    defaults
joblib                    1.0.1              pyhd3eb1b0_0    defaults
jpeg                      9b                   hb83a4c4_2    defaults
jsonschema                3.2.0                      py_2    defaults
jupyter                   1.0.0                    py37_7    defaults
jupyter_client            6.1.12             pyhd3eb1b0_0    defaults
jupyter_console           6.4.0              pyhd3eb1b0_0    defaults
jupyter_core              4.7.1            py37haa95532_0    defaults
jupyterlab_pygments       0.1.2                      py_0    defaults
jupyterlab_widgets        1.0.0              pyhd3eb1b0_1    defaults
kealib                    1.4.14               hde4a422_0    defaults
kiwisolver                1.3.1            py37hd77b12b_0    defaults
krb5                      1.19.2               h5b6d351_0    defaults
libcurl                   7.78.0               h86230a5_0    defaults
libgdal                   3.0.2                ha1b3edf_1    defaults
libiconv                  1.15                 h1df5818_7    defaults
libnetcdf                 4.6.1                hf59b723_4    defaults
libpng                    1.6.37               h2a8f88b_0    defaults
libpq                     12.2                 hb652d5d_1    defaults
libspatialindex           1.9.3                h6c2663c_0    defaults
libspatialite             4.3.0a               h7ffb84d_0    defaults
libssh2                   1.9.0                h7a1dbc1_1    defaults
libtiff                   4.2.0                hd0e1b90_0    defaults
libxml2                   2.9.12               h0ad7f3c_0    defaults
lz4-c                     1.9.3                h2bbff1b_1    defaults
m2w64-expat               2.1.1                         2    defaults
m2w64-gcc-libgfortran     5.3.0                         6    defaults
m2w64-gcc-libs            5.3.0                         7    defaults
m2w64-gcc-libs-core       5.3.0                         7    defaults
m2w64-gettext             0.19.7                        2    defaults
m2w64-gmp                 6.1.0                         2    defaults
m2w64-libiconv            1.14                          6    defaults
m2w64-libwinpthread-git   5.0.0.4634.697f757               2    defaults
m2w64-xz                  5.2.2                         2    defaults
mapclassify               2.4.3              pyhd3eb1b0_0    defaults
markupsafe                2.0.1            py37h2bbff1b_0    defaults
matplotlib-base           3.4.2            py37h49ac443_0    defaults
matplotlib-inline         0.1.2              pyhd3eb1b0_2    defaults
mistune                   0.8.4           py37hfa6e2cd_1001    defaults
mkl                       2021.3.0           haa95532_524    defaults
mkl-service               2.4.0            py37h2bbff1b_0    defaults
mkl_fft                   1.3.0            py37h277e83a_2    defaults
mkl_random                1.2.2            py37hf11a4ad_0    defaults
msys2-conda-epoch         20160418                      1    defaults
munch                     2.5.0                      py_0    defaults
munkres                   1.1.4                      py_0    defaults
nbclient                  0.5.3              pyhd3eb1b0_0    defaults
nbconvert                 6.1.0            py37haa95532_0    defaults
nbformat                  5.1.3              pyhd3eb1b0_0    defaults
nest-asyncio              1.5.1              pyhd3eb1b0_0    defaults
networkx                  2.6.2              pyhd3eb1b0_0    defaults
notebook                  6.4.3            py37haa95532_0    defaults
numexpr                   2.7.3            py37hb80d3ca_1    defaults
numpy                     1.20.3           py37ha4e8547_0    defaults
numpy-base                1.20.3           py37hc2deb75_0    defaults
olefile                   0.46                     py37_0    defaults
openjpeg                  2.4.0                h4fc8c34_0    defaults
openssl                   1.1.1k               h8ffe710_1    conda-forge
osmnx                     1.1.1              pyhd8ed1ab_0    conda-forge
packaging                 21.0               pyhd3eb1b0_0    defaults
pandas                    1.3.2            py37h6214cd6_0    defaults
pandocfilters             1.4.3            py37haa95532_1    defaults
parso                     0.8.2              pyhd3eb1b0_0    defaults
pickleshare               0.7.5           pyhd3eb1b0_1003    defaults
pillow                    8.3.1            py37h4fa10fc_0    defaults
pip                       21.0.1           py37haa95532_0    defaults
postgresql                12.2                 hb652d5d_1    defaults
proj                      6.2.1                h9f7ef89_0    defaults
prometheus_client         0.11.0             pyhd3eb1b0_0    defaults
prompt-toolkit            3.0.17             pyh06a4308_0    defaults
prompt_toolkit            3.0.17               hd3eb1b0_0    defaults
pycparser                 2.20                       py_2    defaults
pygments                  2.10.0             pyhd3eb1b0_0    defaults
pyopenssl                 20.0.1             pyhd3eb1b0_1    defaults
pyparsing                 2.4.7              pyhd3eb1b0_0    defaults
pyproj                    2.6.1.post1      py37h593ac45_1    defaults
pyqt                      5.9.2            py37h6538335_2    defaults
pyrsistent                0.17.3           py37he774522_0    defaults
pysocks                   1.7.1            py37h03978a9_3    conda-forge
python                    3.7.11               h6244533_0    defaults
python-dateutil           2.8.2              pyhd3eb1b0_0    defaults
python_abi                3.7                     2_cp37m    conda-forge
pytz                      2021.1             pyhd3eb1b0_0    defaults
pywin32                   228              py37hbaba5e8_1    defaults
pywinpty                  0.5.7                    py37_0    defaults
pyzmq                     22.2.1           py37hd77b12b_1    defaults
qt                        5.9.7            vc14h73c81de_0    defaults
qtconsole                 5.1.0              pyhd3eb1b0_0    defaults
qtpy                      1.10.0             pyhd3eb1b0_0    defaults
rasterio                  1.1.0            py37h039b02d_0    defaults
requests                  2.26.0             pyhd3eb1b0_0    defaults
rtree                     0.9.7            py37h2eaa2aa_1    defaults
scikit-learn              0.24.2           py37hf11a4ad_1    defaults
scipy                     1.6.2            py37h66253e8_1    defaults
send2trash                1.5.0              pyhd3eb1b0_1    defaults
setuptools                52.0.0           py37haa95532_0    defaults
shapely                   1.7.1            py37h210f175_0    defaults
sip                       4.19.8           py37h6538335_0    defaults
six                       1.16.0             pyhd3eb1b0_0    defaults
snuggs                    1.4.7                      py_0    defaults
sqlite                    3.36.0               h2bbff1b_0    defaults
terminado                 0.9.4            py37haa95532_0    defaults
testpath                  0.5.0              pyhd3eb1b0_0    defaults
threadpoolctl             2.2.0              pyhbf3da8f_0    defaults
tiledb                    2.2.9                hf7ce2e6_0    defaults
tk                        8.6.10               he774522_0    defaults
tornado                   6.1              py37h2bbff1b_0    defaults
traitlets                 5.0.5              pyhd3eb1b0_0    defaults
typing_extensions         3.10.0.0           pyh06a4308_0    defaults
urllib3                   1.26.6             pyhd3eb1b0_1    defaults
vc                        14.2                 h21ff451_1    defaults
vs2015_runtime            14.27.29016          h5e58377_2    defaults
wcwidth                   0.2.5                      py_0    defaults
webencodings              0.5.1                    py37_1    defaults
wheel                     0.37.0             pyhd3eb1b0_0    defaults
widgetsnbextension        3.5.1                    py37_0    defaults
win_inet_pton             1.1.0            py37haa95532_0    defaults
wincertstore              0.2                      py37_0    defaults
winpty                    0.4.3                         4    defaults
xerces-c                  3.2.3                ha925a31_0    defaults
xz                        5.2.5                h62dcd97_0    defaults
zipp                      3.5.0              pyhd3eb1b0_0    defaults
zlib                      1.2.11               h62dcd97_4    defaults
zstd                      1.4.9                h19a0ad4_0    defaults

```

</details>

My temporal fix is to "deactivate" resolving the URL's domain to an IP address by:

```python
import osmnx as ox
# patch resolving the URL's domain to an IP address within `downloader.py`
ox.downloader._config_dns = lambda url: None

lami, lama, lomi, loma = (18.505, 18.550, 73.83, 73.870)
G=ox.graph_from_bbox(lami, lama, lomi, loma, network_type="drive")
```

But I think osmnx should add a fix within `downloader.py`, e.g. in `def _config_dns(url)`:

```python
def _config_dns(url):
    ...
    host = urlparse(url).netloc.split(":")[0]
    try:
        ip = socket.gethostbyname(host)
    except socket.gaierror:
        # fix could go here
        # ugly but working
        ip = host
 
    def _getaddrinfo(*args):
        if args[0] == host:
    ...

```

**Comment by gboeing** [comment]
Thanks for flagging this @jslorrma. Something like your proposed fix (ie, catch the exception and just skip IP address resolution) could work. However, this also means that anyone using OSMnx behind such a proxy might resolve their status endpoint query and their interpreter endpoint query to different Overpass servers given how Overpass's load balancing is configured. This could result in load balancing problems (which possibly may have cropped up in the past). #699 was the best I could come up with for a fix, but I'm open to suggestions. CCing @mmd-osm in case there are any general Overpass load balancing comments/considerations to keep in mind to ensure OSMnx plays nicely with them.

> Pull request #699 breaks any data download in case proxies are used or have to be used, e.g. if one is using osmnx from within a corporate network.

Is this issue really that ubiquitous? It may be, but I'm initially a little skeptical because this feature was released several months ago, and I hadn't heard a report of this until now. I would have thought someone would have hit this problem and reported it by now. Could it possibly be a narrower issue than the scope of any corporate/proxy usage?

**Comment by gboeing** [comment]
FWIW, in the meantime, downgrading to OSMnx v1.1.0 should work as a temporary fix.

**Comment by jslorrma** [comment]
> Is this issue really that ubiquitous? It may be, but I'm initially a little skeptical because this feature was released several months ago, and I hadn't heard a report of this until now. I would have thought someone would have hit this problem and reported it by now. Could it possibly be a narrower issue than the scope of any corporate/proxy usage?

Separate from osmnx, the error `gaierror: [Errno 11001] getaddrinfo failed` seems to bee quite common quite common in the context of the `socket` library (`socket.gethostbyname` or more general `socket.getaddrinfo`) when used from within a corporate / proxied networks, that in addition often use its own DNS servers. Here are some discussions from [stackoverflow.com](https://www.google.com/search?q=gaierror:+%5BErrno+11001%5D+getaddrinfo+failed+site:stackoverflow.com&sxsrf=AOaemvKgrDQVn6FP6Vc4ZfmcR2KBePomOw:1630411579053&sa=X&ved=2ahUKEwiC1Ya4nNvyAhXK6eAKHRroBaAQrQIoBHoECAcQBQ&biw=1920&bih=1089)

In this situation quite often only local name resolution works via the company's DNS Servers and trying to resolve external hosts using utilities, like e.g. `nslookup` fails . Trying e.g. to resolve `overpass-api.de` returns `Non-existent domain` (Windows) and `** server can't find overpass-api.de: NXDOMAIN` (Linux). The python `socket`-library uses the same low-level (OS specific) C-functions like e.g `nslookup` and thus fails, too.

> Thanks for flagging this @jslorrma. Something like your proposed fix (ie, catch the exception and just skip IP address resolution) could work. However, this also means that anyone using OSMnx behind such a proxy might resolve their status endpoint query and their interpreter endpoint query to different Overpass servers given how Overpass's load balancing is configured. This could result in load balancing problems (which possibly may have cropped up in the past). #699 was the best I could come up with for a fix, but I'm open to suggestions. CCing @mmd-osm in case there are any general Overpass load balancing comments/considerations to keep in mind to ensure OSMnx plays nicely with them.

My proposal is to use a public DNS-service, like e.g. the one from [google](https://developers.google.com/speed/public-dns), to resolve a URL to its IP address, if the current approach using `socket` raises an error. Here is an code-example using the google's public [JSON API for DNS over HTTPS](https://developers.google.com/speed/public-dns/docs/doh/json) to receive the IP address from a URL:
```python
import requests

def gethostbyname(host: str, debug: bool = False) -> str:
    # alternatively one can use: https://8.8.8.8/resolve?name={host}
    _dns_url = f"https://dns.google/resolve?name={host}" 
    # send get request
    _response = requests.get(_dns_url)
    
    # print the full response for debugging
    if debug:
        import json
        print(json.dumps(_response.json(), indent=2))
    
    # if request was successful return the host IP(s)
    # Status = 0 -> // NOERROR - Standard DNS response code (32 bit integer)
    if _response.ok and _response.json()["Status"] == 0:
        return [answer["data"] for answer in _response.json()["Answer"]]
    # in case host could not be resolved return the host itself
    else:
        print(f"'{host}' could not be resolved. Response status: {_response.json()['Status']}")
        return host

gethostbyname("overpass-api.de", debug=True)
```
that prints the DNS answer:
```
{'name': 'overpass-api.de.', 'type': 1, 'TTL': 20078, 'data': '178.63.11.215'}
{'name': 'overpass-api.de.', 'type': 1, 'TTL': 20078, 'data': '178.63.48.217'}
```

and returns the list of public IPs for `overpass-api.de` (comparable to the discussion here #698):
```
['178.63.11.215', '178.63.48.217']
```

**Comment by gboeing** [comment]
Interesting. I think this seems like a good idea: try the current approach, and if we hit an error try to resolve the IP using Google instead, then proceed with the rest of the current process.

**Comment by gboeing** [comment]
@jslorrma given your nice diagnostic work on this so far, would you like to open a PR to contribute a fix?

**Comment by gboeing** [comment]
@jslorrma just checking back. Given the work you put into diagnosing this, I'll defer to you if you'd like to make this small contribution yourself. If not, let me know and I can make the proposed changes.

**Comment by jslorrma** [comment]
> @jslorrma just checking back. Given the work you put into diagnosing this, I'll defer to you if you'd like to make this small contribution yourself. If not, let me know and I can make the proposed changes.

Sorry for my not replying. Been pretty busy at work lately.

@gboeing First of all, thank you for asking me to contribute to this beautiful project.  have a couple of spare days and I will be glad to create a pull request on this issue. I haven't done that on GitHub yet, but that shouldn't be a problem.

**Comment by gboeing** [comment]
Thanks @jslorrma sounds good. I'll review the PR when you've opened it.

**Comment by gboeing** [comment]
Proposed fix in #774

**Comment by gboeing** [comment]
Resolved by #774
